Create `marko-web-recaptcha` package and use in contact-us, inquiry & theme-monorail packages

B77Mills opened the pull request:
Add new package with three primary utilities:
 - Loading of the script
 - Generation of the token
 - Validation of the token
 
 Update the Contact Us, Inquiry & ThemeMonorail packages to all use this new reCAPTCHA package to consistently handle reCAPTCHA.  This mean that on submit it will load the script, then generate the token and submit it.  On the server side when handling the submit it uses the validate token function.  
 
Also added the omeda_newsletter routes and config setup to the marko-web-omeda package.  This will then need to be added at the site or shared routes level of the site setup.  Wanted to move it to a core package so it being setup in ever site/theme differently.  

@TODO: Update any sites that are utilizing Contact Us & Inquiry with the newest version and update env variables. The bolded items will need to be done in the related org repos  
 - AB Media
 **-- [] convert currently site/pacakges using recaptcha(ContactUs)**
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher
 
 - ACBM
 **-- [] convert current site/pacakges using recaptcha(ContactUs)**
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher
 
 - Allured 
 **-- [] convert current site/pacakges using recaptcha(ContactUs)**
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher
 
 - Ascend
 **-- [] convert current site/pacakges using recaptcha(ContactUs)**
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher
 
 - Bizbash
 **-- [] convert current site/pacakges using recaptcha(Inquiry)**
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher

 - Cox Matthews
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher
 
 - INDM
 **-- [] convert current site/pacakges using recaptcha(ContactUs)**
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher

 - PMMI
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher
 
 - Rogue
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher
 
  - RR
 -- [X] Create/Confirm recaptchaV3 is setup for site
 -- [X] Add recaptchaV3 env variables in rancher

B77Mills committed:
create new dynamic recatcha script loader

Add new dynamic loading of recatcha v3 script on submit of the newsletter signup forms.

B77Mills committed:
Fix syntax.

B77Mills committed:
Create starting point for recaptcha package

Make new pacakge for the handeling of recaptcha(load, getToken & hopefully token checking)

B77Mills committed:
add window document globals

B77Mills committed:
switch newsletter to use new marko-web-recaptcha package

B77Mills committed:
restructure new recaptcha package

split package into browser and server parts

B77Mills committed:
update newsletter brwoser componets with new recaptcha calls

B77Mills committed:
add omeda newsletter routes to the omeda package

Add omeda newsletter signup routes to the omeda package along with supporting files like the recaptcha config setup.

B77Mills committed:
remove recapthca script component

The new way of including recaptcha will replace this

B77Mills committed:
fix package.jscon for omeda package

B77Mills committed:
udpate validate-token to support passing of actions

Add passing of actions to be able to set both on create and validate of tokens

B77Mills committed:
Update Inquiry Package to use new recaptcha package

Remove vue-recaptcha & update to use new recaptcha package that uses google recaptcha v3.  This mean updating the inclusion of recaptcha script to on submit long with update how the token is generated and validated.

B77Mills committed:
add newseletter config to example site

B77Mills committed:
update default site to call content page with newsletter menu and Inquiry form

B77Mills committed:
fix package.json error

B77Mills committed:
add recaptchaV3 env variable to example website

B77Mills committed:
remove console.log

B77Mills committed:
Reenable sending of emails

B77Mills committed:
set recaptcha as a global var in example website

B77Mills committed:
Update ContactUs package to use new recaptcha package

remove vue-recaptcha & add support for dynamically loading recaptcha on submit and update get token and token validation to use new recaptcha package.

B77Mills committed:
add contact us route & components to content route

B77Mills committed:
remove vue recaptcha

B77Mills committed:
remove old recaptcha env

B77Mills committed:
set actions for omedaNewsletter signups

B77Mills committed:
ensure submitActions is array & score is set.

set defalts for submitActions to emtpy array & score to 0.5 is not passed

B77Mills committed:
remove console.log

B77Mills commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Consolidate reCAPTCHA script in Monorail, contact us & inquiry forms.](https://trello.com/c/v7Jvf8te/97-consolidate-recaptcha-script-in-monorail-contact-us-inquiry-forms)

zarathustra323 commented on docker-compose.yml:
```diff
@@ -333,6 +333,10 @@ services:
       OMEDA_APP_ID: ${EXAMPLE_OMEDA_APP_ID}
       OMEDA_INPUT_ID: ${EXAMPLE_OMEDA_INPUT_ID}
       OMEDA_RAPID_IDENT_PRODUCT_ID: ${EXAMPLE_OMEDA_RAPID_IDENT_PRODUCT_ID-374}
+      RECAPTCHA_V3_SECRET_KEY: ${RECAPTCHA_V3_SECRET_KEY-}
```
Prefix the right-side values with `EXAMPLE_` (like the above)... e.g. `RECAPTCHA_V3_SECRET_KEY: ${EXAMPLE_RECAPTCHA_V3_SECRET_KEY-}`

zarathustra323 commented on packages/marko-web-recaptcha/README.md:
```diff
@@ -0,0 +1,2 @@
+# BaseCMS Object Path
```
Update this file to reflect recaptcha, not object path :)

zarathustra323 commented on packages/marko-web-recaptcha/browser/.eslintrc.js:
```diff
@@ -0,0 +1,21 @@
+    browser: true,
+  },
+  rules: {
+    'vue/max-attributes-per-line': ['error', {
```
Remove this

zarathustra323 commented on packages/marko-web-recaptcha/browser/.eslintrc.js:
```diff
@@ -0,0 +1,21 @@
+module.exports = {
+  extends: [
+    'airbnb-base',
+    'plugin:vue/recommended',
```
Remove this

zarathustra323 commented on packages/marko-web-recaptcha/browser/get-token.js:
```diff
@@ -0,0 +1,6 @@
+export default ({
```
```js
export default ({ siteKey, action }) => {
  if (!siteKey) throw new Error('The recaptcha site key is required');
  if (!action) throw new Error('The recaptcha action is required');
  return window.grecaptcha.execute(siteKey, { action });
}
```

zarathustra323 commented on packages/marko-web-recaptcha/src/validate-token.js:
```diff
@@ -0,0 +1,20 @@
+  minimumScore,
+}) => {
+  const score = minimumScore || 0.5;
+  const submitActions = actions || [];
```
`if (!actions || !actions.length) throw new Error('At least one recaptcha action is required')`

zarathustra323 replied:
Then remove `submitActions` and use `actions` directly

zarathustra323 commented on packages/marko-web-recaptcha/src/validate-token.js:
```diff
@@ -0,0 +1,20 @@
+  actions,
+  minimumScore,
+}) => {
+  const score = minimumScore || 0.5;
```
Also throw error if token or secretKey are missing

zarathustra323 commented on packages/marko-web-recaptcha/src/validate-token.js:
```diff
@@ -0,0 +1,20 @@
+  params.append('secret', secretKey);
+  const recaptchaRes = await fetch('https://www.google.com/recaptcha/api/siteverify', { method: 'post', mode: 'no-cors', body: params });
+  const recaptchaData = await recaptchaRes.json();
+  if (!recaptchaData.success || !submitActions.includes(recaptchaData.action) || recaptchaData.score < score) throw createError(400, 'Unable to validate request because reCAPTCHA failed.');
```
Throw individual error for each error condition

zarathustra323 commented on packages/marko-web-recaptcha/browser/load.js:
```diff
@@ -0,0 +1,32 @@
+}));
+
+export default async ({ siteKey }) => {
+  if (!recaptchaLoadPromise) recaptchaLoadPromise = recaptchaLoad(siteKey);
```
`if (!window.recaptchaLoadPromise) window.recaptchaLoadPromise = recaptchaLoad(siteKey);`

zarathustra323 commented on packages/marko-web-recaptcha/browser/load.js:
```diff
@@ -0,0 +1,32 @@
+let recaptchaLoadPromise;
```
Delete this since the `window` global will be used

zarathustra323 committed:
Adjust script loading to force on load

zarathustra323 committed:
Move recaptcha loading to `created` hook

B77Mills committed:
Update setting and errors on required props

B77Mills committed:
move recaptcha loading/error state check to button disable

B77Mills committed:
move recaptchaLoad to `create()`

Move the inclusion of the reCAPTCHA script to create() call of inquiry, contact us & newsletter sign components.

zarathustra323 commented on packages/marko-web-contact-us/browser/form.vue:
```diff
@@ -84,45 +84,44 @@
         >
           Hold up, we're processing your submission...
         </p>
+        <p
```
Move this to the top of the row and change to `v-if` instead of `v-else-if`

zarathustra323 replied:
And use `A recaptcha error occurred` instead of the errorLabel

zarathustra323 commented on packages/marko-web-inquiry/browser/default-form.vue:
```diff
@@ -150,16 +150,15 @@
       </div>
     </div>
+    <pre v-if="recaptcha.error" class="alert alert-danger text-danger">
+      An error occurred: {{ recaptcha.error.message }}
```
Change to `A recaptcha error occurred`

zarathustra323 commented on packages/marko-web-inquiry/routers/submit.js:
```diff
@@ -30,33 +30,11 @@ module.exports = ({ queryFragment, notification, confirmation }) => asyncRoute(a
-    return true;
-  };
-
-  if (!payload || payload.email === '') {
```
Add this back but change to:

```js
if (!payload || !payload.email) {
  throw exception('Invalid form submission');
}
```

zarathustra323 commented on packages/marko-web-inquiry/routers/submit.js:
```diff
@@ -30,33 +30,11 @@ module.exports = ({ queryFragment, notification, confirmation }) => asyncRoute(a
 
-  const { error } = console;
-
-  const exception = (message, code = 400) => {
```
Restore this

zarathustra323 commented on packages/marko-web-recaptcha/src/validate-token.js:
```diff
@@ -0,0 +1,19 @@
+  params.append('secret', secretKey);
+  const recaptchaRes = await fetch('https://www.google.com/recaptcha/api/siteverify', { method: 'post', mode: 'no-cors', body: params });
+  const recaptchaData = await recaptchaRes.json();
+  if (!recaptchaData.success || !actions.includes(recaptchaData.action) || recaptchaData.score < minimumScore) throw createError(400, 'Unable to validate request because reCAPTCHA failed.');
```
Throw for each error type

zarathustra323 commented on services/example-website/browser/index.js:
```diff
@@ -1,11 +1,17 @@
 import Browser from '@parameter1/base-cms-marko-web/browser';
-import OmedaIdentityX from '@parameter1/base-cms-marko-web-omeda-identity-x/browser';
-import GTM from '@parameter1/base-cms-marko-web-gtm/browser';
+// import OmedaIdentityX from '@parameter1/base-cms-marko-web-omeda-identity-x/browser';
```
Delete the commented out code

B77Mills committed:
Prefix env var with EXAMPLE_

B77Mills committed:
update readme

B77Mills committed:
update eslintrc

 - remove extends: plugin:vue/recommended & rules vue/max-attributes-pre-line

B77Mills committed:
Add error handeling for unset props

B77Mills committed:
update actions check and error message

B77Mills committed:
Update recaptcha validation error handeling

Split recaptchaData checks into 3 separate checks(sucess, actionCheck & score). Throw `failed`, `action is not supported` &  `score did not meet minimum requirements.`

B77Mills committed:
Split recaptcha error handeling out of if submitted check

B77Mills committed:
Update recaptcha error label

B77Mills committed:
Add payload.email check & exception support back

B77Mills committed:
remove commented out stuff

B77Mills committed:
Update recaptcha error label

B77Mills committed:
Add EXAMPLE to send grid env var
